metals/src/main/scala/scala/meta/internal/metals/ConnectionProvider.scala

```diff
@@ -0,0 +1,729 @@
+            val message =
+              "Failed to connect with build server, no functionality will work."
+            val details = " See logs for more details."
+            languageClient.showMessage(
+              new MessageParams(MessageType.Error, message + details)
+            )
+            scribe.error(message, e)
+            BuildChange.Failed
+          }
+          .flatMap(compileAllOpenFiles)
+          .map { res =>
+            buildServerPromise.trySuccess(())
+            res
+          }
+      }
+    }
+
+    private object GenerateBspConfigAndConnectProvider
+        extends ConnectProvider[
+          (BuildServerProvider, Boolean),
+          GenerateBspConfigAndConnect,
+        ] {
+
+      def run(config: (BuildServerProvider, Boolean)): Future[BuildChange] = {
+        val (buildTool, shutdownServer) = config
+        tables.buildTool.chooseBuildTool(buildTool.executableName)
+        maybeChooseServer(buildTool.buildServerName, alreadySelected = false)
+        for {
+          _ <-
+            if (shutdownServer) DisconnectProvider.run(shutdownServer)
+            else Future.unit
+          status <- buildTool
+            .generateBspConfig(
+              folder,
+              args => bspConfigGenerator.runUnconditionally(buildTool, args),
+              statusBar,
+            )
+          shouldConnect = handleGenerationStatus(buildTool, status)
+          status <-
+            if (shouldConnect) CreateSessionProvider.run(false)
+            else Future.successful(BuildChange.Failed)
+        } yield status
+      }
+
+      /**
+       * Handles showing the user what they need to know after an attempt to
+       * generate a bsp config has happened.
+       */
+      private def handleGenerationStatus(
+          buildTool: BuildServerProvider,
+          status: BspConfigGenerationStatus,
+      ): Boolean = status match {
+        case bsp.BspConfigGenerationStatus.Generated =>
+          tables.buildServers.chooseServer(buildTool.buildServerName)
+          true
+        case bsp.BspConfigGenerationStatus.Cancelled => false
+        case bsp.BspConfigGenerationStatus.Failed(exit) =>
+          exit match {
+            case Left(exitCode) =>
+              scribe.error(
+                s"Creation of .bsp/${buildTool.buildServerName} failed with exit code: $exitCode"
+              )
+              languageClient.showMessage(
+                Messages.BspProvider.genericUnableToCreateConfig
+              )
+            case Right(message) =>
+              languageClient.showMessage(
+                Messages.BspProvider.unableToCreateConfigFromMessage(
+                  message
+                )
+              )
+          }
+          false
+      }
+    }
+
+    val isImportInProcess = new AtomicBoolean(false)
+
+    private object BloopInstallAndConnectProvider
+        extends ConnectProvider[
+          (BloopInstallProvider, String, Boolean, Boolean),
```

Comment:
```suggestion
          BloopInstallConfig,
```
Should we have a class here?